File: dynamic_connectivity/quick_union.py
""" QuickUnion is a lazy approach.

*Limitations:*

- Trees can get too tall
- Find too expensive (could be N array accesses)
"""


class QuickUnion:
    """ data[i] is parent of i. Root of i is id[id[...id[i]...]] """

    def __init__(self, n):
        """ Initializing list of size n where value is same as index
        Here it means that each node is a root of it's own tree
        Time Complexity: O(n)

        :param n: number of elements
        """
        self.data = []
        for i in range(n):
            self.data.append(i)

    def root(self, elem):
        """ Finding the root of element

        :param elem: element of which root is needed
        :return: root of elem
        """
        while elem != self.data[elem]:
            elem = self.data[elem]
        return elem

    def connected(self, elem1, elem2):
        """ elem1 and elem2 are connected iff they have same root

        :param elem1: element 1
        :param elem2: element 2
        :return: returns true iff two elem1 and elem2  are connected, else
            false
        :rtype: bool
        """
        return self.root(elem1) == self.root(elem2)

    def union(self, elem1, elem2):
        """ To merge components containing elem1 and elem2, set the id of
        elem1's root to the id of elem2's root

        :param elem1: element 1
        :param elem2: element 2
        """
        root_elem1 = self.root(elem1)
        root_elem2 = self.root(elem2)
        self.data[root_elem1] = root_elem2


def main():
    """ operational function """
    maze = QuickUnion(10)
    maze.union(4, 3)
    maze.union(3, 8)
    maze.union(6, 5)
    maze.union(9, 4)
    maze.union(2, 1)
    print("is 0-7 connected: ", maze.connected(0, 7))
    print("is 8-9 connected: ", maze.connected(8, 9))
    maze.union(5, 0)
    maze.union(7, 2)
    maze.union(6, 1)
    maze.union(1, 0)
    print("is 0-7 connected: ", maze.connected(0, 7))


if __name__ == "__main__":
    main()

File: _other_languages/cpp/DynamicConnectivity/QuickUnion.cpp
/**
 * Dynamic Connectivity
 *
 * Given a set of N objects
 * - Union Command: connect two objects.
 * - Find/connected query: is there a path connecting the two objects?
 *
 * Quick-Find and Quick-Union are only going to tell us if there is a path
 * between two objects. It is not going to give us the complete path.
 *
 * Connected components: Maximal set of objects that are  mutually connected.
*/
#include <iostream>

/**
 * QuickUnion (Lazy Approach)
 * Data Structure: Integer array id[] of size N
 * Interpretation: id[i] is parent of i
 * Root/parent of i is id[id[id[...id[i]...]]]
 *
 * QuickUnion defect:
 * - Trees can get too tall
 * - Find too expensive (could be N array accesses)
*/
class QuickUnion {
private:
    int size;
    int* id;

    /**
     * Finding the root of element
     * @param elem element of which root is needed
     * @return root of elem
    */
    int root(int elem) {
        while (elem != id[elem])
            elem = id[elem];
        return elem;
    }

public:
    /**
     * Initializing list of size n where value is same as index
     * Time Complexity: O(n)
     * @param n number of elements
    */
    QuickUnion(int n): size(n) {
        id = new int[n];
        for (int i = 0; i < size; i++)
            id[i] = i;
    }

    /**
     * p and q are connected iff they have same root
     * @param p element 1
     * @param q element 2
     * @return true iff two p and q are connected, else false
    */
    bool connected(int p, int q) {
        return root(p) == root(q);
    }

    /**
     * To merge components containing p and q, set the id of p's
     * root to the id of q's root
     * @param p element 1
     * @param q element 2
    */
    void joinUnion(int p, int q) {
        int i = root(p);
        int j = root(q);
        id[i] = j;
    }
};

/* Main Driver Function */
int main()
{
    QuickUnion quickUnion(10);
    quickUnion.joinUnion(4, 3);
    quickUnion.joinUnion(3, 8);
    quickUnion.joinUnion(6, 5);
    quickUnion.joinUnion(9, 4);
    quickUnion.joinUnion(2, 1);
    std::cout << "is 0 connected to 7? " << quickUnion.connected(0, 7) << "\n";
    std::cout << "is 8 connected to 9? " << quickUnion.connected(8, 9) << "\n";
    quickUnion.joinUnion(5, 0);
    quickUnion.joinUnion(7, 2);
    quickUnion.joinUnion(6, 1);
    quickUnion.joinUnion(1, 0);
    std::cout << "is 0 connected to 7? " << quickUnion.connected(0, 7) << "\n";
    return 0;
}